### PR TITLE
Move all precision transformation into optimizeFunction()

### PR DESCRIPTION
**Pull request by jfix71**
Summary: Move quantization/profiling/fp16 conversion into `optimizeFunction()`. This means we can call lower just once in a normal way, instead of forcing it to occur twice. Also cleans up callers to `compile()` so they do not need to worry about when/how to quantize/profile/convert to fp16.

Documentation: Added.

Test Plan: All tests pass.

Fixes https://github.com/pytorch/glow/issues/2648

### CODE REVIEW

File: include/glow/Backends/CompilationOptions.h [hunk 1]
```diff
@@ -33,6 +35,30 @@ struct CompilationOptions {
 
   /// Insert TraceEvents between all instructions for profiling.
   bool autoInstrument{false};
+
+  // Different precision modes. Note that convertToFP16 and quantize can be
```
Comment (jackm321) [inline include/glow/Backends/CompilationOptions.h]: doxygen here and other places

File: include/glow/Backends/CompilationOptions.h [hunk 2]
```diff
@@ -33,6 +35,30 @@ struct CompilationOptions {
...
+  // specified simultaneously.
+  bool convertToFP16{false};
+  bool quantize{false};
+  bool profileForQuantization{false};
```
Comment (jackm321) [inline include/glow/Backends/CompilationOptions.h]: quantize and profileForQuantization should be mutually exclusive right

File: lib/Backends/Backend.cpp [hunk 3]
```diff
@@ -15,23 +15,75 @@
...
-  ::glow::lower(F, /* loweredMap */ nullptr, this);
+  // Lower the graph into a sequence of low-level linear algebra operations. If
+  // profiling, pass a nullptr for the backend, as all nodes should be lowered.
+  ::glow::lower(F, &opts.loweredInfoMap,
+                opts.profileForQuantization ? nullptr : this,
+                opts.precisionModeKindSet);
+
+  // Transform given precision mode; may quantize, convert to fp16, or
+  // instrument with profiling nodes. This must be done after lowering.
+  transformForPrecisionMode(*this, F, opts);
```
Comment (jackm321) [inline lib/Backends/Backend.cpp]: Should transformForPrecisionMode be moved to global scope like the rest of the function here, there's no real reason to have it be in Backend.cpp right?

File: include/glow/Backends/CompilationOptions.h (hunk 2 above)
Comment (jfix71) [reply include/glow/Backends/CompilationOptions.h]: Correct -- I suppose I could replace them with an enum for QuantizationMode. I wanted to still allow fp16 conversion + quantization.

File: lib/Backends/Backend.cpp (hunk 3 above)
Comment (jfix71) [reply lib/Backends/Backend.cpp]: I guess if we want to unit test it then making it global would make sense; otherwise it's basically just a helper that was extracted to make the logic in `optimizeFunction()` more readable.

Additionally, I think it along with `optimizeFunction()` should be moved to GraphOptimizer.cpp (see https://github.com/pytorch/glow/pull/2703).

File: include/glow/Backends/Backend.h [hunk 4]
```diff
@@ -102,7 +102,7 @@ class Backend {
   virtual bool shouldShareBuffers() const { return true; }
 
   /// Optimize the Function \p F given compilation options \p opts.
-  void optimizeFunction(Function *F, const CompilationOptions &opts) const;
+  void optimizeFunction(Function *&F, CompilationOptions &opts) const;
```
Comment (nadavrot) [inline include/glow/Backends/Backend.h]: I don't think that passing a pointer to a reference is a very good interface.

File: include/glow/ExecutionEngine/ExecutionEngine.h [hunk 5]
```diff
@@ -118,8 +118,8 @@ class ExecutionEngine final {
   /// Make \p networkName the function name for
   /// the entry point of the network and prepend all generated
   /// files with this name.
-  void save(Function *F, const CompilationOptions &opts,
-            llvm::StringRef outputDir, llvm::StringRef networkName);
+  void save(Function *&F, CompilationOptions &opts, llvm::StringRef outputDir,
```
Comment (nadavrot) [inline include/glow/ExecutionEngine/ExecutionEngine.h]: Why did you remove the const? Are we modifying the compilation options inside? This will make it really difficult to track things around.

File: lib/Backends/Backend.cpp [hunk 6]
```diff
@@ -15,23 +15,75 @@
...
 #include "glow/Graph/Graph.h"
 #include "glow/Graph/PlaceholderBindings.h"
 #include "glow/IR/Instrs.h"
 #include "glow/Optimizer/Optimizer.h"
 
 using namespace glow;
 
-void Backend::optimizeFunction(Function *F,
-                               const CompilationOptions &opts) const {
+static void transformForPrecisionMode(const Backend &B, Function *&F,
```
Comment (nadavrot) [inline lib/Backends/Backend.cpp]: Why do we need to pass the backend around? This really complicates things. We should not continue to increase the number of parematers to all of our optimizations functions. This is a sign that we don't have clean APIs.

File: lib/Backends/CMakeLists.txt [hunk 7]
```diff
@@ -38,7 +38,8 @@ target_link_libraries(Backend
                         CodeGen
                         Graph
                         IR
-                        Optimizer)
+                        Optimizer
```
Comment (nadavrot) [inline lib/Backends/CMakeLists.txt]: Why would the backends, who's job is to generate code, depend on the quantizer? The quantizer is just a procedure that mutates the graph.

**Comment by nadavrot** [comment]
Jordan, this PR adds new parameters to functions, and new dependencies to libraries. This is an indication that the design becomes more complex and the interfaces less clean. Can we simplify things somehow?

File: include/glow/ExecutionEngine/ExecutionEngine.h (hunk 5 above)
Comment (jfix71) [reply include/glow/ExecutionEngine/ExecutionEngine.h]: The only thing that's being modified is the `LoweredInfoMap` -- I could move that to be a pointer and keep `CompilationOptions` `const`. Then the `LoweredInfoMap` would need to be explicitly kept in the caller's scope or higher.

File: lib/Backends/Backend.cpp (hunk 6 above)
Comment (jfix71) [reply lib/Backends/Backend.cpp]: The Backend is required in `optimizeFunction()` (and then into here) because it reports on what the specific Backend supports + what should be lowered (via `isOpSupported()` + `shouldLower()`), and it also allows the Backend to transform the graph (via `transformPostLowering()`).

It's passed in here because the quantizer only quantizes Nodes given what the Backend supports (via `isOpSupported()`). It's just a helper method here for `optimizeFunction()`.

One way we could consider redesigning this would be to refactor Backend into multiple classes, like a BackendPreferences class. But I don't see a way to not pass at least something for the backend in here right now.

File: lib/Backends/CMakeLists.txt (hunk 7 above)
Comment (jfix71) [reply lib/Backends/CMakeLists.txt]: Agreed, this was because the `optimizeFunction()` lives in the Backend class, which I don't think makes sense anyway. I put up a PR to move `optimizeFunction()` to Optimizer in https://github.com/pytorch/glow/pull/2703.

**Comment by jfix71** [comment]
I've cleaned this up a bit here and through other PRs:
- `CompilationOptions` is again `const` when passed into `compile()` and beyond.
- `optimizeFunction()` is now in `Optimizer`, which is what now depends on `Quantization`/`Converter`.
- `compile()` returns the transformed `Function` (i.e. transformed after all passes for folding, optimizations, lowering, and precision transformation), instead of passing `F` into `compile()` as a reference to a pointer.

Some other modifications that might be a good idea (whether for this PR or not), as discussed offline with @opti-mix and @jackm321:
- Perhaps `compile()` should not return the transformed `F` -- one alternative here is to keep a pointer to `F` in `CompilationOptions` and update it, though this means a big refactor to set `F` in `opts` everywhere instead of passing it into `compile()`. 
    - Note that the current returning of `F` mirrors `quantizeFunction()` and `profileQuantization()` which also return the transformed `F`.
- Rename `CompilationOptions` as it now includes things like `PlaceholderBindings`, `KindSet`, `LoweredInfoMap`, etc. (perhaps `CompilationContext` or `CompilationConfiguration`?).

If anyone has further/other thoughts/opinions here, happy to hear them.

**Comment by jfix71** [comment]
I've given this an initial refactoring on the `Loader`, `fr2en`, and `compareAgainstInterpreter()`:
- `CompilationContext` replaces `CompilationOptions`. It's a struct containing:
    - `CompilationMode` enum (`Infer` or `Train`)
    - `PlaceholderBindings` pointer for this compilation
    - `LoweredInfoMap` pointer for this compilation
    - `PrecisionConfiguration`, a struct containing:
        - `convertToFP16` bool flag
        - `QuantizationMode` (enum of `Quantize`/`Profile`/`None`)
        - `QuantizationConfiguration` (from #2751), used if `QuantizationMode::Quantize`
    - `BackendOptions`, a struct passed into `Backend::compile()` after `optimizeFunction()` is called, containing:
        - `collectConstants` bool flag
        - `autoInstrument` bool flag

RE: `compile()` returning the transformed `Function` -- I think it's the best option for now. It mirrors the `quantizeFunction()` and `profileQuantization()` calls which also return the Function. Some other options which I find less desirable: 
- Pass `F` as a pointer to a pointer and update `F` inside `compile()`/its callees -- this is similar to the initial version of a reference to a pointer. Don't really love this as an interface.
- Keep a pointer to the final Function in the `CompilationContext` -- don't love this either. Would require removing `const` when passing `CompilationContext` around.

File: include/glow/Optimizer/Optimizer.h [hunk 8]
```diff
@@ -75,9 +75,9 @@ Function *profileQuantization(PlaceholderBindings &bindings, Function *F,
 std::unique_ptr<IRFunction> generateAndOptimizeIR(Function *F, const Backend &B,
                                                   bool shouldShareBuffers);
 
-/// Optimize the Function \p F given compilation options \p opts for Backend \B.
-void optimizeFunction(Function *F, const Backend &B,
-                      const CompilationOptions &opts);
+/// Optimize the Function \p F given compilation context \p cctx for Backend \B.
```
Comment (nadavrot) [inline include/glow/Optimizer/Optimizer.h]: //  \returns ... 

It is not clear from the signature if the function is cloned and optimized or just modified in place.

**Comment by nadavrot** [comment]
@jfix71 Thanks for doing this work. The changes to introduce CompilationContext look great. 

It is not clear from the API if we return a cloned function (like the quantizer does) or the original function. I don't see why we would like to return the original function, because we had just passed that in. On the other hand, I am not seeing a call to clone(). I think that cloning the function may block some optimizations because constants will have users from multiple functions and we will run into a problem where we can't delete the original function and need to keep the new function. On the other hand it's kind of strange that the optimizer changes the function inside compile().

**Comment by jfix71** [comment]
>It is not clear from the API if we return a cloned function (like the quantizer does) or the original function. I don't see why we would like to return the original function, because we had just passed that in. On the other hand, I am not seeing a call to clone(). I think that cloning the function may block some optimizations because constants will have users from multiple functions and we will run into a problem where we can't delete the original function and need to keep the new function. On the other hand it's kind of strange that the optimizer changes the function inside compile().

@nadavrot It returns a clone if we profile/quantize, and otherwise returns the original function. Our APIs are inconsistent here -- clone for profile/quantize, do not clone for optimize/convertToFP16. I actually think we should remove cloning from all of them -- I think if the caller wants to keep a clone of the original Function it should be responsible for doing so, and understand the Function passed in will be mutated. That way `compile()` would still return nothing, and all our interfaces would be consistent. WDYT?

> To avoid merge conflicts is there a way to start landing the changes to CompilationContext that are ready?

Yeah I can move `CompilationContext` to a separate PR. I was putting it in here because many of the changes for `CompilationContext` will touch the same places this PR will, but it would probably be too big a PR to reasonably review.

**Comment by nadavrot** [comment]
@jfix71 Removing cloning makes sense to me. The overall plan looks good. Let me know if you want me to stamp this PR or if you want to land the pieces in multiple PRs.

**Comment by jfix71** [comment]
@nadavrot I'll do it in a few PRs. First one is up in #2792. Will file another one later removing cloning from quantize/profile.

**Comment by jfix71** [comment]
OK, I think this is finally ready! Would love for some reviews so I can land this and stop rebasing it 😃

File: docs/GraphOptimizationPipeline.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  - `Profile`: Add special profiling nodes to the graph to gather a histogram of
+    values flowing through each Node in the graph. This can then be used to
+    automatically quantize the graph.
+
+  - `Quantize`: Quantize the graph, given a previously gathered profile as
+    mentioned above. This converts all Float inputs and outputs of each Node if
+    the backend supports the node as quantized.
+
+  - `ConvertToFP16`: Convert all Float inputs and outputs of Nodes to
+    Float16. Note that this does require any sort of profile. It can also be
```
Comment (gcatron) [inline docs/GraphOptimizationPipeline.md]: this does *not* require

File: lib/Optimizer/GraphOptimizer.cpp [hunk 10]
```diff
@@ -2720,16 +2722,75 @@ static llvm::Error checkAllNodesSupported(const Function &F, const Backend &B) {
...
+
+  if (precConfig.convertToFP16) {
+    TypeAToTypeBFunctionConverter converter(*F, ElemKind::FloatTy,
+                                            ElemKind::Float16Ty,
+                                            &precConfig.precisionModeKindSet);
+    converter.convert();
+  }
+}
+
 llvm::Error glow::optimizeFunction(Function *F, const Backend &B,
```
Comment (jackm321) [inline lib/Optimizer/GraphOptimizer.cpp]: Given that the documentation is very specific about these optimizations, maybe add a note here to update the documentation when this function changes

File: include/glow/Optimizer/CompilationContext.h [hunk 11]
```diff
@@ -19,20 +19,90 @@
...
+    Profile,  /// Add profiling nodes for quantization statistics gathering.
+  } quantMode{QuantizationMode::None};
+
+  /// Configuration for Quantization.
+  quantization::QuantizationConfiguration quantConfig;
+
+  /// Whether to convert the FloatTy to Float16Ty in the Function.
+  bool convertToFP16{false};
+
+  /// Used during Quantization, Profiling, and convertToFP16.
```
Comment (jackm321) [inline include/glow/Optimizer/CompilationContext.h]: Since this comment could be more helpful, probably mentioning something about this being kinds to exclude or something like that

File: lib/Onnxifi/InlineOnnxifi.cpp [hunk 12]
```diff
@@ -57,20 +57,16 @@ InlineGraph::initGraph(const void *onnxModel, size_t onnxModelSize,
   computeModelHash(onnxModel, onnxModelSize, modelHash_);
   optimize(function_, CompilationMode::Infer);
 
-  // -- Profile --
```
Comment (jackm321) [inline lib/Onnxifi/InlineOnnxifi.cpp]: woo, glad this is gone!
